public/app/features/trails/otel/util.ts

```diff
@@ -312,16 +294,334 @@ export async function updateOtelJoinWithGroupLeft(trail: DataTrail, metric: stri
+      // add the dep env to var filters if not present and isPromoted
+      const depEnvFromVarFilters = prevVarFilters.filter((f) => f.key === 'deployment_environment');
+
+      // if promoted and no dep env has been chosen yet, set the default
+      if (isPromoted && depEnvFromVarFilters.length === 0) {
+        prevVarFilters = [...prevVarFilters, ...defaultDepEnvFilter];
+      }
+
+      prevVarFilters = [...prevVarFilters, ...urlVarFilters];
+
+      filtersVariable.setState({
+        filters: prevVarFilters,
+        hide: VariableHide.hideVariable,
+      });
+    }
+  }
+  // 1. Get the otel join query for state and variable
+  // Because we need to define the deployment environment variable
+  // we also need to update the otel join query state and variable
+  const resourcesObject: OtelResourcesObject = getOtelResourcesObject(trail);
+  // THIS ASSUMES THAT WE ALWAYS HAVE DEPLOYMENT ENVIRONMENT!
+  // FIX THIS SO THAT WE HAVE SOME QUERY EVEN IF THERE ARE NO OTEL FILTERS
+  const otelJoinQuery = getOtelJoinQuery(resourcesObject);
+
+  // update the otel join query variable too
+  otelJoinQueryVariable.setState({ value: otelJoinQuery });
+
+  // 2. Update state with the following
+  // - otel join query
+  // - otelTargets used to filter metrics
+  // now we can filter target_info targets by deployment_environment="somevalue"
+  // and use these new targets to reduce the metrics
+  // for initialization we also update the following
+  // - has otel resources flag
+  // - and default to useOtelExperience
+  const otelTargets = await totalOtelResources(datasourceUid, timeRange, resourcesObject.filters);
+
+  // we pass in deploymentEnvironments and hasOtelResources on start
+  // RETHINK We may be able to get rid of this check
+  // a non standard data source is more missing job and instance matchers
+  if (hasOtelResources && deploymentEnvironments && !initialOtelCheckComplete) {
+    trail.setState({
+      otelTargets,
+      otelJoinQuery,
+      hasOtelResources,
+      // Previously checking standardization for having deployment environments
+      // Now we check that there are target_info labels that are not promoted
+      isStandardOtel: (nonPromotedOtelResources ?? []).length > 0,
+      useOtelExperience: isEnabledInLocalStorage,
+      nonPromotedOtelResources,
+      initialOtelCheckComplete: true,
+      resettingOtel: false,
+      afterFirstOtelCheck: true,
+      isUpdatingOtel: false,
+    });
+  } else {
+    // we are updating on variable changes
+    trail.setState({
+      otelTargets,
+      otelJoinQuery,
+      resettingOtel: false,
+      afterFirstOtelCheck: true,
+      isUpdatingOtel: false,
+    });
+  }
+}
+
+function checkLabelPromotion(filters: AdHocVariableFilter[], nonPromotedOtelResources: string[] = []) {
+  const nonPromoted = filters.filter((f) => nonPromotedOtelResources.includes(f.key));
+  const promoted = filters.filter((f) => !nonPromotedOtelResources.includes(f.key));
```

Comment:
Similarly to feedback in `public/app/features/trails/otel/api.ts`, we can avoid expensive `Array.prototype.includes` calls in a loop by using `Set.prototype.has`:

```suggestion
  const nonPromotedResources = new Set(nonPromotedOtelResources);
  const nonPromoted = filters.filter((f) => nonPromotedResources.has(f.key));
  const promoted = filters.filter((f) => !nonPromotedResources.has(f.key));
```